### PR TITLE
explicit types between chain processors and provider message assembly

### PR DESCRIPTION
**Pull request by agouin**
Phase 1 of #825 

Modifies retention of messages in the `IBCMessagesCache` to explicit types `provider.PacketInfo`, `provider.ConnectionInfo`, and `provider.ChannelInfo`. This removes complexity from the `ChainProcessor` as it can blindly forward the observed IBC message properties without having to worry about constructing the start of the next message in the sequence. The provider methods are instead responsible for that. `CosmosChainProcessor` now constructs these new public types in the event parser code and the message handlers can slim down to single handlers for the different message types.

Optimizes retries to retry message assembly after 1 block, and message sending after 2 blocks. This prevents double spending and leads to clean logs for all of the channel handshake, connection handshake, and packet flow sequences.

### CODE REVIEW

File: relayer/chains/cosmos/event_parser.go
Comment (mark-rushakoff) [inline relayer/chains/cosmos/event_parser.go:236]: It's okay to have omitted these two cases from the updated code?

File: relayer/chains/cosmos/event_parser.go
Comment (mark-rushakoff) [inline relayer/chains/cosmos/event_parser.go]: I don't see why the `parseXYZ` functions have moved away from methods... well, I do, because the `packetInfo` type has moved to a new package.

A method is sensible here because this is modifying `res` in-place. Changing `res` from the receiver to a parameter hurts readability.

It looks like these could all be of form `(res *PacketInfo) parseAttributes(log *zap.Logger, attrs []sdk.Attribute)`, `(res *ChannelInfo) parseAttributes(attrs []sdk.Attribute)`. If they consistently accept `log` as the first parameter, then we have a consistent interface like

```go
type messageInfo interface { // Find a better name?
  parseAttrs(log *zap.Logger, attrs []sdk.Attribute) // Attrs shorthand is good for an unexported interface
}
```

And then hopefully above, we can avoid `var info any` in favor of `var info provider.messageInfo` for a bit stronger typing. (I don't like the name `messageInfo`, but we can find something more descriptive.)

File: relayer/chains/cosmos/message_handlers.go
Comment (mark-rushakoff) [inline relayer/chains/cosmos/message_handlers.go:12]: Since there is no default case in this switch, a comment would help to explain why these are the only actions we handle or why we don't log unexpected types.

File: relayer/chains/cosmos/message_handlers.go
Comment (mark-rushakoff) [inline relayer/chains/cosmos/message_handlers.go]: `ibcMessagesCache` is too long and just repeats the already long type name. Just `c` or `cache` will do here.

File: relayer/chains/cosmos/message_handlers.go
Comment (mark-rushakoff) [inline relayer/chains/cosmos/message_handlers.go:58]: There is enough slice manipulation here, that it wouldn't hurt to add

```go
if !ccp.log.Core().Enabled(zapcore.DebugLevel) {
  return
}
```

to avoid allocating in the common, not-debug-enabled case.

File: relayer/chains/cosmos/types.go
Comment (mark-rushakoff) [inline relayer/chains/cosmos/types.go]: Would love to get this away from `any` if we can find the right interface.

File: relayer/chains/cosmos/types.go
Comment (mark-rushakoff) [inline relayer/chains/cosmos/types.go]: Could this method just be `update` since its receiver is a `lastestClientState`?

File: relayer/processor/types.go
Comment (mark-rushakoff) [inline relayer/processor/types.go]: Since this is used repeatedly as an argument to a logging call that may not be enabled, I would prefer this to defer the `strings.Split` until we are actually going to log it.

In this case it should be simpler to rely on `zap.Stringer` rather than an object marshaler.

```go
type ShortAction string

func (a ShortAction) String() string {
  	split := strings.Split(action, ".")
	return split[len(split)-1]
}
```

And then instead of `zap.String("msg", ShortAction(a))`, it's `zap.Stringer("msg", ShortAction(a))` (`Stringer` instead of `String` to lazily evaluate the `.String()` method call.

File: relayer/provider/provider.go
Comment (mark-rushakoff) [inline relayer/provider/provider.go]: Call these fields `DestPort` and `DestChannel`, as that is noticeably shorter and just as clear.

File: relayer/provider/provider.go
Comment (mark-rushakoff) [inline relayer/provider/provider.go]: This could be `Ack`.

File: relayer/provider/provider.go
Comment (mark-rushakoff) [inline relayer/provider/provider.go]: `ConnID` and `CounterpartyConnID` will be clear enough here.

File: relayer/provider/provider.go
Comment (mark-rushakoff) [inline relayer/provider/provider.go]: `CounterpartyConnID`

File: relayer/provider/cosmos/tx.go
Comment (mark-rushakoff) [inline relayer/provider/cosmos/tx.go]: This translation of a `provider.PacketInfo` occurs 4 times in this file. Extract a function like `toCosmosPacket(pi provider.PacketInfo) chantypes.Packet` so that we don't risk an inconsistency among the 4 occurrences.

In this case a function makes more sense than a method, as we want to keep the cosmos `chantypes` package decoupled from the `provider` package.

File: relayer/processor/path_end_runtime.go
Comment (mark-rushakoff) [inline relayer/processor/path_end_runtime.go]: Use `>=` here -- better to not assume that you'll land exactly on the max count, in the event of a programmer error where two increments happen before this check.

File: relayer/processor/path_end_runtime.go
Comment (mark-rushakoff) [inline relayer/processor/path_end_runtime.go]: `Cache` is redundant on the field names. `packetProcessing, connProcessing, channelProcessing` is sufficiently clear.

File: relayer/processor/path_end_runtime.go
Comment (mark-rushakoff) [inline relayer/processor/path_end_runtime.go]: `>=`

**Review by mark-rushakoff — COMMENTED** [review]
I've left a few comments. Renaming fields and types could happen separately, but the more structural types of changes would be better to include in this PR before merging.

File: relayer/chains/cosmos/event_parser_test.go
Comment (jtieri) [inline relayer/chains/cosmos/event_parser_test.go]: imo it was easier to quickly understand what's happening here when the function was returning a value instead of mutating the argument passed to it.

File: relayer/processor/types_internal.go
Comment (jtieri) [inline relayer/processor/types_internal.go]: `deleteMessages` probably suffices here since the receiver has cache in the type name so you can assume it's deleting messages from a cache.

Same goes for all the `deleteCachedMessages` methods in this file

File: relayer/chains/cosmos/event_parser.go
Comment (agouin) [reply relayer/chains/cosmos/event_parser.go:236]: yes, they are unused for packet message assembly

File: relayer/processor/types.go
Comment (jtieri) [inline relayer/processor/types.go]: This feels like we should never be hitting the panic here unless there is a programmer error but I could be misconstruing the code path to get here.

Could we just log an error and backup through the call stack to the main query loop?

**Review by jtieri — COMMENTED** [review]
Looking good so far. 

I think this opens the door for a lot of general cleanup, especially around the `Provider` interface.

File: relayer/chains/cosmos/event_parser.go
Comment (agouin) [reply relayer/chains/cosmos/event_parser.go]: Since `PacketInfo` etc. are non-cosmos-specific types, I think I will do this with an aliased type inside the cosmos chain processor package.

Comment (agouin) [reply relayer/chains/cosmos/event_parser.go]: Added types for this: `packetInfo`, `connectionInfo`, `channelInfo` which alias the public provider types and have the `parseAttrs` methods, all internal to the `CosmosChainProcessor`.

File: relayer/chains/cosmos/message_handlers.go
Comment (agouin) [reply relayer/chains/cosmos/message_handlers.go:12]: done

File: relayer/chains/cosmos/message_handlers.go
Comment (agouin) [reply relayer/chains/cosmos/message_handlers.go]: done

File: relayer/chains/cosmos/message_handlers.go
Comment (agouin) [reply relayer/chains/cosmos/message_handlers.go:58]: good call, done

File: relayer/chains/cosmos/types.go
Comment (agouin) [reply relayer/chains/cosmos/types.go]: `ibcMessageInfo` interface added

File: relayer/chains/cosmos/types.go
Comment (agouin) [reply relayer/chains/cosmos/types.go]: done

File: relayer/processor/types.go
Comment (agouin) [reply relayer/processor/types.go]: nice optimization, done

File: relayer/provider/provider.go
Comment (agouin) [reply relayer/provider/provider.go]: done

File: relayer/provider/provider.go
Comment (agouin) [reply relayer/provider/provider.go]: done

File: relayer/provider/provider.go
Comment (agouin) [reply relayer/provider/provider.go]: done

File: relayer/provider/provider.go
Comment (agouin) [reply relayer/provider/provider.go]: done

File: relayer/provider/cosmos/tx.go
Comment (agouin) [reply relayer/provider/cosmos/tx.go]: done

File: relayer/processor/path_end_runtime.go
Comment (agouin) [reply relayer/processor/path_end_runtime.go]: done

File: relayer/processor/path_end_runtime.go
Comment (agouin) [reply relayer/processor/path_end_runtime.go]: done

File: relayer/processor/path_end_runtime.go
Comment (agouin) [reply relayer/processor/path_end_runtime.go]: done

File: relayer/chains/cosmos/event_parser_test.go
Comment (agouin) [reply relayer/chains/cosmos/event_parser_test.go]: modified this to be methods on top of alias types

File: relayer/processor/types_internal.go
Comment (agouin) [reply relayer/processor/types_internal.go]: done

File: relayer/processor/types.go
Comment (agouin) [reply relayer/processor/types.go]: done

File: relayer/chains/cosmos/types.go
Comment (mark-rushakoff) [inline relayer/chains/cosmos/types.go]: It's awkward to have a type declared far away from its method definitions; and I would further argue that `types.go` is an anti-pattern (presumably cargo-culted from other packages we depend on) that we should discontinue, on the basis of resemblance to a [utils package](https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common).

The `...Info` types could move into `event_parser.go` near the method definitions, or perhaps these types and their methods could move into a new `info.go` file.

File: relayer/chains/cosmos/types.go
Comment (mark-rushakoff) [inline relayer/chains/cosmos/types.go]: GitHub makes it difficult to search the content of a PR once you've collapsed some of the files -- if this satisfies an interface we can keep it, otherwise it should be removed.

**Review by mark-rushakoff — APPROVED** [review]
Two more small comments that are not blocking merge at this point.

File: relayer/chains/cosmos/types.go
Comment (agouin) [reply relayer/chains/cosmos/types.go]: yes forgot to remove this, done.

File: relayer/chains/cosmos/types.go
Comment (agouin) [reply relayer/chains/cosmos/types.go]: done

File: relayer/processor/path_end_runtime.go
Comment (danwt) [inline relayer/processor/path_end_runtime.go:310]: hi @agouin @mark-rushakoff @jtieri 
Please could you tell me the reason for this check?
Thank you

**Pull request merged**
